feat: trace all previous blockhashes

letypequividelespoubelles opened the pull request:
<!-- CURSOR_SUMMARY -->
> [!NOTE]
> Plumbs historical block-hash collection into tracing and replay: gather prior hashes, add last-block hash, and wire through BlockCapturer, ZkTracer/Hub, RPC, and tests with EIP-2935/4788 touches.
> 
> - **Blockhash tracing & replay**:
>   - Replace `BlockHashReaper` with map-based storage in `Reaper`; add consistency checks; include EIP-2935/4788 system-tx touches.
>   - Overhaul `module/blockhash` to track first/last blocks, inject historical hashes, add parent-hash op, auto-emit missing historical calls, and include last-block placeholder; `BlockhashOperation` now uses `Hash`.
>   - `ConflationSnapshot`: new `from(...)`, `historicalBlockHashes()`, refined `toBlockHashLookup`.
> - **Tracer/Hub plumbing**:
>   - `ZkTracer` and all `*Hub` constructors accept historical hashes and pass to `Blockhash`.
>   - `Fork.getForkFromBesuBlockchainService` overload for `BlockchainService`.
>   - `ZkCounter`: count blockhash lines at conflation start; fix counting loops and report unchecked modules as 0.
> - **Capture/RPC**:
>   - `CaptureToFile`/`GenerateConflatedTracesV2`: fetch fork and historical block hashes from `BlockchainService` and pass to `BlockCapturer`/`ZkTracer`.
>   - `BlockCapturer`: accept historical hashes, record end-block hash, and simplify per-opcode handling.
> - **Snapshots**:
>   - `BlockHeaderSnapshot`: store `parentBeaconBlockRoot` as optional hex string with null-guard.
> - **Tests & utilities**:
>   - Replay and toy environments parse conflations first, derive and pass historical hashes; add Prague Sepolia tests; adjust defaults to Prague; minor disables.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 077d7429d23093008359979a6270437aa5c702e6. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

letypequividelespoubelles committed:
feat: trace all previous blockhashes

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
feat: some stuff

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
Merge branch 'arith-dev' into 2348-add-previous-blockhashes-to-the-conflation

letypequividelespoubelles committed:
rebase

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
rm the try catch, if else is better + fix when dealing with first blocks of the chain

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
fix line counting

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
spotless

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
update corset

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
ras

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
rebase

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
ras

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
ras rebase

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
add blockhash line count at begining of the conflation in zktracer

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
test: move blockhash test to besu node tests

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

amkCha commented on testing/src/main/java/net/consensys/linea/testing/ToyExecutionEnvironmentV2.java:
```diff
@@ -90,7 +91,7 @@ public static ToyExecutionEnvironmentV2.ToyExecutionEnvironmentV2Builder builder
         .unitTestsChain(chainConfig)
         .testInfo(testInfo)
-        .tracer(new ZkTracer(chainConfig));
+        .tracer(new ZkTracer(chainConfig, null));
```
Is it possible to make another constructor with a different signature ? So that we don't pass null everywhere, I think it's possible

letypequividelespoubelles replied:
yes, sure, good idea.

letypequividelespoubelles committed:
fix: wrong cast to short

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
rebase

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
ras rebase

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
ras

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
default constructor wo blockchainservices

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
redo from scratch cleaner way

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
make it cleaner

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
add last block too

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
Merge branch 'arith-dev' into 2348-add-previous-blockhashes-to-the-conflation

letypequividelespoubelles committed:
fix: trm leading zeros ...

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
ras stupid stuff

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
some more stupid stuff

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
fix: line count before end conflation

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

DavePearce reviewed (approved):
LGTM

amkCha commented on arithmetization/src/main/java/net/consensys/linea/plugins/rpc/tracegeneration/GenerateConflatedTracesV2.java:
```diff
@@ -112,16 +115,20 @@ private TraceFile generateTraceFile(PluginRpcRequest request) {
+      final BlockchainService blockchain =
+          BesuServiceProvider.getBesuService(besuContext, BlockchainService.class);
       // Retrieve fork from Besu plugin API with block number
       final Fork fork = getForkFromBesuBlockchainService(besuContext, fromBlock, toBlock);
```
Is it possible to change the signature of `getForkFromBesuBlockchainService` to get the blockchain service directly ? Because in the method, I recreate a service from the context, so it's redundant :)

amkCha replied:
Like  `getForkFromBesuBlockchainService(blockchainService, fromBlock, toBlock);`

letypequividelespoubelles replied:
yep

amkCha commented on arithmetization/src/main/java/net/consensys/linea/plugins/rpc/tracegeneration/GenerateConflatedTracesV2.java:
```diff
@@ -112,16 +115,20 @@ private TraceFile generateTraceFile(PluginRpcRequest request) {
     if (cachedTraceFileAvailable(path)) {
       log.info("[TRACING] cached trace for {}-{} detected as {}", fromBlock, toBlock, path);
     } else {
+      final BlockchainService blockchain =
```
can we call it `blockchainService` or `bcService` if too long ?

amkCha commented on testing/src/main/java/net/consensys/linea/testing/ToyExecutionEnvironmentV2.java:
```diff
@@ -257,7 +258,7 @@ public GeneralStateTestCaseEipSpec buildGeneralStateTestCaseSpec(ProtocolSpec pr
     final BlockHeader blockHeader =
         ExecutionEnvironment.getLineaBlockHeaderBuilder(Optional.empty())
-            .number(DEFAULT_BLOCK_NUMBER)
+            .number(firstBlockNumber)
```
nice !

amkCha commented on arithmetization/src/main/java/net/consensys/linea/plugins/rpc/tracegeneration/GenerateConflatedTracesV2.java:
```diff
@@ -85,7 +88,7 @@ public TraceFile execute(final PluginRpcRequest request) {
 
   private TraceFile generateTraceFile(PluginRpcRequest request) {
-    Stopwatch sw = Stopwatch.createStarted();
+    final Stopwatch sw = Stopwatch.createStarted();
```
just checking if it's on purpose or added it for debugging ?

letypequividelespoubelles replied:
I think I added it for debugging, but then as it's not modified, it's a small improvemen on the fly. better put final when it's needed to be sure we don't modify by mistake

amkCha commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/blockhash/Blockhash.java:
```diff
@@ -35,33 +41,40 @@
-  public Blockhash(Hub hub, Wcp wcp) {
+  public Blockhash(Hub hub, Wcp wcp, Map<Long, Hash> historicalBlockHashes) {
+    blockHashMap = historicalBlockHashes;
+    tracingInProd = !historicalBlockHashes.isEmpty();
```
I think the tracer should always be for prod, and we need to adapt our test setup to fit the test case. What we can do is feed the tracer with a list of mock hashes in the constructor, that should have a length of 256 ?

letypequividelespoubelles replied:
hum ... let's discuss this.
In theory I do agree

amkCha commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/blockhash/Blockhash.java:
```diff
@@ -131,12 +193,60 @@ public int spillage(Trace trace) {
+              .orElseThrow(
+                  () ->
+                      new IllegalArgumentException(
+                          "Block not found for number: " + blockNumberAttempt))
```
To be sure we target the error message right, can we add sth like `To retrieve hashes, block not found for ..."

amkCha commented on arithmetization/src/main/java/net/consensys/linea/zktracer/ZkTracer.java:
```diff
@@ -77,22 +94,30 @@ public ZkTracer(
   }
 
+  public ZkTracer(ChainConfig chain) {
+    this(chain, new HashMap<>());
```
Here, we can create a dummy list i/o an empty one. From what I understand, as it's not used in the constraints, it shouldn't fail

letypequividelespoubelles replied:
it'll fail if we have a BLOCKHASH opcode the evm runs, that's why it's not so trivial to feed it with a non trivial map here

letypequividelespoubelles committed:
ras rebase

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
amelie's comment

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
feat: store last block of the conflation blockhash

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
feat: modify the replay mechanism to dd the blockhash history

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
Merge branch 'arith-dev' into 2348-add-previous-blockhashes-to-the-conflation

letypequividelespoubelles committed:
update consraints

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
trigger blockhash from eip2935 section

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
ras

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
fix: don't call blockhash if exceptional

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
retrieve historical blockhaseh for replay tests

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
test: skip tests where we can't extract data from gson

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
tmp: gradle properties

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
add tests for sepolia replay preparation

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
Merge branch 'arith-dev' into 2348-add-previous-blockhashes-to-the-conflation

letypequividelespoubelles committed:
add system tx for block capturer

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
ras

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

DavePearce committed:
Encode/decode parentBeaconBlockRoot as hex string

This changes the type of the field parentBeaconBlockRoot in
BlockHeaderSnapshot to be an (optional) String instead of a Bytes32
instance.  This is to avoid decoding issues because the GSONDecoder is
not aware of the Bytes32 type.

DavePearce committed:
don't load clique headers from Shanghai onwards

The clique consensus protocol was only used prior to shanghai, hence we
cannot extract key information from block headers generated from
Shanghai nodes and onwards.

DavePearce committed:
restore BlockHashSnapshot

This restores this component, whilst retaining flexibility around the
use of Map<Long,Hash>.  The goal is simply to ensure all block hashes
are stored in the JSON file as hex strings.

DavePearce committed:
minor fixes

letypequividelespoubelles committed:
add replay tests and little debug

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
ras

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
rebase

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

DavePearce committed:
fix Reaper and ConflationSnapshot

This puts in fixes for the methods Reaper.touchBlockHash() and also
ConflationSnapshot.from(), and applies spotless.

DavePearce committed:
add replay tests from PRAGUE

letypequividelespoubelles committed:
rebase

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/blockcapture/reapers/Reaper.java:
```diff
@@ -75,6 +80,38 @@ public void enterBlock(
+      final UInt256 timestamp = UInt256.valueOf(header.getTimestamp());
+      final UInt256 keyTimestamp = timestamp.mod(HISTORY_BUFFER_LENGTH);
+      conflationStorage.touch(EIP4788_BEACONROOT_ADDRESS, timestamp);
+      conflationStorage.touch(EIP4788_BEACONROOT_ADDRESS, keyTimestamp);
```
The keys that are touched by 4788 system transactions are
```
brKey := timestamp mod 8191
tsKey := brKey + 8191
```
Am I misunderstanding here ?

letypequividelespoubelles replied:
nope, you're right

letypequividelespoubelles committed:
rebase

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
nothing important

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

OlivierBBB commented on arithmetization/src/test/java/net/consensys/linea/replaytests/ReplayTests.java:
```diff
@@ -187,6 +189,7 @@ void incorrectCreationCapture(TestInfo testInfo) {
     replay(OLD_MAINNET_TESTCONFIG, "4323985.mainnet.json.gz", testInfo);
   }
 
+  @Disabled("Fails to create the ConflationSnapshot from the gson file")
```
Should this read `json` ? (Several cases in this file, but I also see `gson` in other places)

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/blockhash/Blockhash.java:
```diff
@@ -73,26 +83,69 @@ public void traceStartBlock(
+    checkBlockHashConsistancies(e);
+  }
+
+  private void checkBlockHashConsistancies(BlockhashOperation op) {
```
Typo `checkBlockHashConsistencies`

OlivierBBB reviewed (commented):
Not much of a review. One potential issue in the logic wrt the keys at which the `EIP4788_BEACONROOT_ADDRESS` contract's storage gets probed.
